db/database_test.go

```diff
@@ -1907,3 +1908,179 @@ func TestSyncFnMutateBody(t *testing.T) {
 	log.Printf("rev: %s", rev.BodyBytes)
 
 }
+
+// Multiple clients are attempting to push the same new revision concurrently; first writer should be successful,
+// subsequent writers should fail on CAS, and then identify that revision already exists on retry.
```

Comment:
Have you been able to verify while running this and the tests below that these are triggering CAS retry (as opposed to having the first goroutine finish updating before the others initially load the doc)?  I'd expect this to be the case given the way in which the goroutines are started, but it would be useful to have some sort of validation that this test is exercising the intended code path.

Reply:
Yeah, I was able to verify this; all tests are triggering CAS retry whilst running. Added debug level logging in `WriteUpdate` method on `CouchbaseBucketGoCB` to make sure the behavior. All the tests are revised to make use of leaky bucket to inject callback in query invocation. Here is the sample test results:

```
=== RUN   **TestConcurrentPushDifferentUpdateNonWinningRevision**
2020-01-03T13:54:41.241-08:00 [INF] db.TestConcurrentPushDifferentUpdateNonWinningRevision: Setup logging: level: debug - keys: CRUD
2020-01-03T13:54:41.245-08:00 [INF] Successfully opened bucket test_data_bucket
2020-01-03T13:54:41.248-08:00 [INF] Set query timeouts for bucket test_data_bucket to cluster:1m15s, bucket:1m15s
2020-01-03T13:54:41.248-08:00 [INF] Flushing bucket test_data_bucket
2020-01-03T13:54:42.199-08:00 [DBG] RetryLoop retrying Wait until bucket has 0 items after flush after 100 ms.
2020-01-03T13:54:42.306-08:00 [DBG] RetryLoop retrying Wait until bucket has 0 items after flush after 200 ms.
2020-01-03T13:54:42.512-08:00 [DBG] RetryLoop retrying Wait until bucket has 0 items after flush after 400 ms.
2020-01-03T13:54:42.918-08:00 [DBG] RetryLoop retrying Wait until bucket has 0 items after flush after 800 ms.
2020-01-03T13:54:43.723-08:00 [DBG] RetryLoop retrying Wait until bucket has 0 items after flush after 1600 ms.
2020-01-03T13:54:45.329-08:00 [INF] GoCBCustomSGTranscoder Opening Couchbase database test_data_bucket on <http://localhost:8091> as user "test_data_bucket"
2020-01-03T13:54:45.331-08:00 [INF] Successfully opened bucket test_data_bucket
2020-01-03T13:54:45.334-08:00 [INF] Set query timeouts for bucket test_data_bucket to cluster:1m15s, bucket:1m15s
2020-01-03T13:54:45.509-08:00 [INF] Design docs successfully created for view version 2.1.
2020-01-03T13:54:45.509-08:00 [INF] Initializing indexes with numReplicas: 0...
2020-01-03T13:54:45.614-08:00 [INF] Verifying index availability for bucket test_data_bucket...
2020-01-03T13:54:45.618-08:00 [INF] Indexes ready for bucket test_data_bucket.
2020-01-03T13:54:45.620-08:00 [INF] WaitForIndexEmpty found 0 results.  GSI index appears to be empty.
2020-01-03T13:54:45.620-08:00 [INF] Created background task: "CleanAgedItems" with interval 1m0s
2020-01-03T13:54:45.620-08:00 [INF] Created background task: "InsertPendingEntries" with interval 2.5s
2020-01-03T13:54:45.620-08:00 [INF] Created background task: "CleanSkippedSequenceQueue" with interval 30m0s
2020-01-03T13:54:45.624-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 1-a
2020-01-03T13:54:45.624-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"" (0 bytes, ttl:0)
2020-01-03T13:54:45.624-08:00 [DBG] CRUD+: Saving doc (seq: #1, id: doc1 rev: 1-a)
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Stored doc "doc1" / "1-a" as #1
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 2-a
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"1-a" (26 bytes, ttl:0)
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Increased sequence batch to 2
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Saving doc (seq: #2, id: doc1 rev: 2-a)
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Stored doc "doc1" / "2-a" as #2
2020-01-03T13:54:45.625-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 3-a
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"2-a" (25 bytes, ttl:0)
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Saving doc (seq: #3, id: doc1 rev: 3-a)
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Stored doc "doc1" / "3-a" as #3
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 4-a
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"3-a" (26 bytes, ttl:0)
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Increased sequence batch to 4
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Saving doc (seq: #4, id: doc1 rev: 4-a)
2020-01-03T13:54:45.626-08:00 [DBG] CRUD+: Stored doc "doc1" / "4-a" as #4
2020-01-03T13:54:45.627-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 2-b
2020-01-03T13:54:45.627-08:00 [DBG] CRUD+: Got old revision "doc1" / "1-a" --> 26 bytes
2020-01-03T13:54:45.627-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"" (0 bytes, ttl:0)
2020-01-03T13:54:45.627-08:00 [DBG] CRUD+: updateDoc("doc1"): Rev "2-b" leaves "4-a" still current
2020-01-03T13:54:45.627-08:00 [DBG] CRUD+: Saving doc (seq: #5, id: doc1 rev: 4-a)
2020-01-03T13:54:45.627-08:00 [DBG] CRUD+: Stored doc "doc1" / "2-b" as #5
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 3-b2
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"2-b" (24 bytes, ttl:0)
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: updateDoc("doc1"): Rev "3-b2" leaves "4-a" still current
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: Saving doc (seq: #6, id: doc1 rev: 4-a)
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 3-b1
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"2-b" (24 bytes, ttl:0)
2020-01-03T13:54:45.628-08:00 [DBG] CRUD+: updateDoc("doc1"): Rev "3-b1" leaves "4-a" still current
2020-01-03T13:54:45.629-08:00 [DBG] CRUD+: Saving doc (seq: #7, id: doc1 rev: 4-a)
2020-01-03T13:54:45.629-08:00 [DBG] CRUD+: Stored doc "doc1" / "3-b1" as #7
2020-01-03T13:54:45.629-08:00 [DBG] CRUD+: CAS WriteUpdate RetryLoop retrying for doc "doc1", attempt 1
2020-01-03T13:54:45.629-08:00 [DBG] CRUD+: Invoking sync on doc "doc1" rev 3-b2
2020-01-03T13:54:45.629-08:00 [DBG] CRUD+: Got old revision "doc1" / "2-b" --> 24 bytes
2020-01-03T13:54:45.630-08:00 [DBG] CRUD+: Backed up revision body "doc1"/"" (0 bytes, ttl:0)
2020-01-03T13:54:45.630-08:00 [DBG] CRUD+: Increased sequence batch to 8
2020-01-03T13:54:45.630-08:00 [DBG] CRUD+: updateDoc("doc1"): Rev "3-b2" leaves "4-a" still current
2020-01-03T13:54:45.630-08:00 [DBG] CRUD+: Saving doc (seq: #8, id: doc1 rev: 4-a)
2020-01-03T13:54:45.630-08:00 [DBG] CRUD+: Stored doc "doc1" / "3-b2" as #8
2020-01-03T13:54:45.631-08:00 [DBG] CRUD+: Released unused sequences #9-#15
2020-01-03T13:54:45.631-08:00 [DBG] Terminating background task: "CleanAgedItems"
2020-01-03T13:54:45.631-08:00 [DBG] Terminating background task: "InsertPendingEntries"
2020-01-03T13:54:45.631-08:00 [DBG] Terminating background task: "CleanSkippedSequenceQueue"
2020-01-03T13:54:45.631-08:00 [INF] db.TestConcurrentPushDifferentUpdateNonWinningRevision: Reset logging
--- PASS: TestConcurrentPushDifferentUpdateNonWinningRevision (4.39s)
PASS
```